hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/WriteStatus.java

```diff
@@ -87,54 +92,91 @@ public WriteStatus() {
     if (trackSuccessRecords) {
-      writtenRecords.add(record);
+      writtenRecordDelegates.add(HoodieRecordDelegate.fromHoodieRecord(record));
     }
+    updateStatsForSuccess(optionalRecordMetadata);
+  }
+
+  /**
+   * Used by native write handles like HoodieRowCreateHandle and HoodieRowDataCreateHandle.
+   *
+   * @see WriteStatus#markSuccess(HoodieRecord, Option)
+   */
+  @PublicAPIMethod(maturity = ApiMaturityLevel.EVOLVING)
+  public void markSuccess(HoodieRecordDelegate recordDelegate, Option<Map<String, String>> optionalRecordMetadata) {
+    if (trackSuccessRecords) {
+      writtenRecordDelegates.add(Objects.requireNonNull(recordDelegate));
+    }
+    updateStatsForSuccess(optionalRecordMetadata);
+  }
+
+  private void updateStatsForSuccess(Option<Map<String, String>> optionalRecordMetadata) {
     totalRecords++;
 
     // get the min and max event time for calculating latency and freshness
-    if (optionalRecordMetadata.isPresent()) {
-      String eventTimeVal = optionalRecordMetadata.get().getOrDefault(METADATA_EVENT_TIME_KEY, null);
-      try {
-        if (!StringUtils.isNullOrEmpty(eventTimeVal)) {
-          int length = eventTimeVal.length();
-          long millisEventTime;
-          // eventTimeVal in seconds unit
-          if (length == 10) {
-            millisEventTime = Long.parseLong(eventTimeVal) * 1000;
-          } else if (length == 13) {
-            // eventTimeVal in millis unit
-            millisEventTime = Long.parseLong(eventTimeVal);
-          } else {
-            throw new IllegalArgumentException("not support event_time format:" + eventTimeVal);
-          }
-          long eventTime = DateTimeUtils.parseDateTime(Long.toString(millisEventTime)).toEpochMilli();
-          stat.setMinEventTime(eventTime);
-          stat.setMaxEventTime(eventTime);
-        }
-      } catch (DateTimeException | IllegalArgumentException e) {
-        LOG.debug(String.format("Fail to parse event time value: %s", eventTimeVal), e);
+    String eventTimeVal = optionalRecordMetadata.orElse(Collections.emptyMap())
+        .getOrDefault(METADATA_EVENT_TIME_KEY, null);
+    if (isNullOrEmpty(eventTimeVal)) {
+      return;
+    }
+    try {
+      int length = eventTimeVal.length();
+      long millisEventTime;
+      // eventTimeVal in seconds unit
+      if (length == 10) {
+        millisEventTime = Long.parseLong(eventTimeVal) * 1000;
+      } else if (length == 13) {
+        // eventTimeVal in millis unit
+        millisEventTime = Long.parseLong(eventTimeVal);
+      } else {
+        throw new IllegalArgumentException("not support event_time format:" + eventTimeVal);
       }
+      long eventTime = DateTimeUtils.parseDateTime(Long.toString(millisEventTime)).toEpochMilli();
+      stat.setMinEventTime(eventTime);
+      stat.setMaxEventTime(eventTime);
+    } catch (DateTimeException | IllegalArgumentException e) {
+      LOG.debug(String.format("Fail to parse event time value: %s", eventTimeVal), e);
```

Comment:
this diff is due to simplifying the nested condition block with early return; logic remains the same